File: src/main/java/io/github/jukomu/langchain4jcodeplatform/controller/UserController.java
package io.github.jukomu.langchain4jcodeplatform.controller;

import com.mybatisflex.core.paginate.Page;
import io.github.jukomu.langchain4jcodeplatform.annotation.AuthCheck;
import io.github.jukomu.langchain4jcodeplatform.common.BaseResponse;
import io.github.jukomu.langchain4jcodeplatform.common.DeleteRequest;
import io.github.jukomu.langchain4jcodeplatform.exception.ErrorCode;
import io.github.jukomu.langchain4jcodeplatform.model.dto.user.*;
import io.github.jukomu.langchain4jcodeplatform.model.vo.LoginUserVo;
import io.github.jukomu.langchain4jcodeplatform.model.vo.UserVo;
import io.github.jukomu.langchain4jcodeplatform.service.UserService;
import io.github.jukomu.langchain4jcodeplatform.util.ResultUtils;
import io.github.jukomu.langchain4jcodeplatform.util.ThrowUtils;
import jakarta.servlet.http.HttpServletRequest;
import lombok.RequiredArgsConstructor;
import org.springframework.web.bind.annotation.*;

import static io.github.jukomu.langchain4jcodeplatform.constant.UserConstant.ADMIN_ROLE;
import static io.github.jukomu.langchain4jcodeplatform.constant.UserConstant.DEFAULT_ROLE;

/**
 * 用户 控制层。
 *
 * @author JUKOMU
 */
@RestController
@RequestMapping("/user")
@RequiredArgsConstructor
public class UserController {

    private final UserService userService;

    /**
     * 用户注册
     */
    @PostMapping("/register")
    public BaseResponse<Long> register(@RequestBody UserRegisterDto userRegisterDto) {
        ThrowUtils.throwIf(userRegisterDto == null, ErrorCode.PARAMS_ERROR);
        long userRegisterId = userService.register(userRegisterDto);
        return ResultUtils.success(userRegisterId);
    }

    /**
     * 用户登录
     */
    @PostMapping("/login")
    public BaseResponse<LoginUserVo> login(@RequestBody UserLoginDto userLoginDto, HttpServletRequest request) {
        ThrowUtils.throwIf(userLoginDto == null || request == null, ErrorCode.PARAMS_ERROR);
        LoginUserVo login = userService.login(userLoginDto, request);
        return ResultUtils.success(login);
    }

    /**
     * 获取当前登录用户
     */
    @GetMapping("/get/login")
    @AuthCheck(hasRole = DEFAULT_ROLE)
    public BaseResponse<LoginUserVo> getLoginUser(HttpServletRequest request) {
        ThrowUtils.throwIf(request == null, ErrorCode.PARAMS_ERROR);
        LoginUserVo loginUser = userService.getLoginUser(request);
        return ResultUtils.success(loginUser);
    }

    /**
     * 用户注销
     */
    @GetMapping("/logout")
    @AuthCheck(hasRole = DEFAULT_ROLE)
    public BaseResponse<Boolean> logout(HttpServletRequest request) {
        ThrowUtils.throwIf(request == null, ErrorCode.PARAMS_ERROR);
        boolean logout = userService.logout(request);
        return ResultUtils.success(logout);
    }

    /**
     * 获取用户 （管理员）
     */
    @GetMapping("/{id}")
    @AuthCheck(hasRole = ADMIN_ROLE)
    public BaseResponse<UserVo> getUser(@PathVariable long id) {
        ThrowUtils.throwIf(id <= 0, ErrorCode.PARAMS_ERROR);
        UserVo userVo = userService.getUser(id);
        return ResultUtils.success(userVo);
    }

    /**
     * 添加用户 （管理员）
     *
     * @param userAddDto
     * @return
     */
    @PostMapping()
    @AuthCheck(hasRole = ADMIN_ROLE)
    public BaseResponse<Long> addUser(@RequestBody UserAddDto userAddDto) {
        ThrowUtils.throwIf(userAddDto == null, ErrorCode.PARAMS_ERROR);
        long userId = userService.addUser(userAddDto);
        return ResultUtils.success(userId);
    }

    /**
     * 更新用户 （管理员）
     *
     * @param userUpdateDto
     * @return
     */
    @PutMapping
    @AuthCheck(hasRole = ADMIN_ROLE)
    public BaseResponse<Boolean> updateUser(@RequestBody UserUpdateDto userUpdateDto) {
        ThrowUtils.throwIf(userUpdateDto == null || userUpdateDto.getId() == null, ErrorCode.PARAMS_ERROR);
        boolean update = userService.updateUser(userUpdateDto);
        return ResultUtils.success(update);
    }

    /**
     * 删除用户 （管理员）
     *
     * @param deleteRequest
     * @return
     */
    @DeleteMapping
    @AuthCheck(hasRole = ADMIN_ROLE)
    public BaseResponse<Boolean> deleteUser(@RequestBody DeleteRequest deleteRequest) {
        ThrowUtils.throwIf(deleteRequest == null || deleteRequest.getId() == null, ErrorCode.PARAMS_ERROR);
        boolean deleted = userService.deleteUser(deleteRequest);
        return ResultUtils.success(deleted);
    }

    /**
     * 获取用户列表 （管理员）
     *
     * @param userQueryDto
     * @return
     */
    @PostMapping("/list")
    @AuthCheck(hasRole = ADMIN_ROLE)
    public BaseResponse<Page<UserVo>> getUsers(@RequestBody UserQueryDto userQueryDto) {
        ThrowUtils.throwIf(userQueryDto == null, ErrorCode.PARAMS_ERROR);
        Page<UserVo> users = userService.getUsers(userQueryDto);
        return ResultUtils.success(users);
    }

}

File: LangChain4j-Code-Platform-Frontend/src/api/userController.ts
// @ts-ignore
/* eslint-disable */
import request from '@/request'

/** 此处后端没有提供注释 PUT /user */
export async function updateUser(body: API.UserUpdateDto, options?: { [key: string]: any }) {
  return request<API.BaseResponseBoolean>('/user', {
    method: 'PUT',
    headers: {
      'Content-Type': 'application/json',
    },
    data: body,
    ...(options || {}),
  })
}

/** 此处后端没有提供注释 POST /user */
export async function addUser(body: API.UserAddDto, options?: { [key: string]: any }) {
  return request<API.BaseResponseLong>('/user', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    data: body,
    ...(options || {}),
  })
}

/** 此处后端没有提供注释 DELETE /user */
export async function deleteUser(body: API.DeleteRequest, options?: { [key: string]: any }) {
  return request<API.BaseResponseBoolean>('/user', {
    method: 'DELETE',
    headers: {
      'Content-Type': 'application/json',
    },
    data: body,
    ...(options || {}),
  })
}

/** 此处后端没有提供注释 GET /user/${param0} */
export async function getUser(
  // 叠加生成的Param类型 (非body参数swagger默认没有生成对象)
  params: API.getUserParams,
  options?: { [key: string]: any }
) {
  const { id: param0, ...queryParams } = params
  return request<API.BaseResponseUserVo>(`/user/${param0}`, {
    method: 'GET',
    params: { ...queryParams },
    ...(options || {}),
  })
}

/** 此处后端没有提供注释 GET /user/get/login */
export async function getLoginUser(options?: { [key: string]: any }) {
  return request<API.BaseResponseLoginUserVo>('/user/get/login', {
    method: 'GET',
    ...(options || {}),
  })
}

/** 此处后端没有提供注释 POST /user/list */
export async function getUsers(body: API.UserQueryDto, options?: { [key: string]: any }) {
  return request<API.BaseResponsePageUserVo>('/user/list', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    data: body,
    ...(options || {}),
  })
}

/** 此处后端没有提供注释 POST /user/login */
export async function login(body: API.UserLoginDto, options?: { [key: string]: any }) {
  return request<API.BaseResponseLoginUserVo>('/user/login', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    data: body,
    ...(options || {}),
  })
}

/** 此处后端没有提供注释 GET /user/logout */
export async function logout(options?: { [key: string]: any }) {
  return request<API.BaseResponseBoolean>('/user/logout', {
    method: 'GET',
    ...(options || {}),
  })
}

/** 此处后端没有提供注释 POST /user/register */
export async function register(body: API.UserRegisterDto, options?: { [key: string]: any }) {
  return request<API.BaseResponseLong>('/user/register', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    data: body,
    ...(options || {}),
  })
}
